torchmetrics/detection/map_new.py

```diff
@@ -0,0 +1,600 @@
+            )
+
+
+class MAP(Metric):
+    r"""
+    Computes the `Mean-Average-Precision (mAP) and Mean-Average-Recall (mAR)\
+    <https://jonathan-hui.medium.com/map-mean-average-precision-for-object-detection-45c121a31173>`_\
+    for object detection predictions.
+    Optionally, the mAP and mAR values can be calculated per class.
+
+    Predicted boxes and targets have to be in Pascal VOC format
+    (xmin-top left, ymin-top left, xmax-bottom right, ymax-bottom right).
+    See the :meth:`update` method for more information about the input format to this metric.
+
+    For an example on how to use this metric check the `torchmetrics examples\
+    <https://github.com/PyTorchLightning/metrics/blob/master/tm_examples/detection_map.py>`_
+
+    .. note::
+        This metric is a wrapper for the
+        `pycocotools <https://github.com/cocodataset/cocoapi/tree/master/PythonAPI/pycocotools>`_,
+        which is a standard implementation for the mAP metric for object detection. Using this metric
+        therefore requires you to have `pycocotools` installed. Please install with ``pip install pycocotools`` or
+        ``pip install torchmetrics[detection]``.
+
+    .. note::
+        This metric requires you to have `torchvision` version 0.8.0 or newer installed (with corresponding
+        version 1.7.0 of torch or newer). Please install with ``pip install torchvision`` or
+        ``pip install torchmetrics[detection]``.
+
+    .. note::
+        As the pycocotools library cannot deal with tensors directly, all results have to be transfered
+        to the CPU, this might have an performance impact on your training.
+
+    Args:
+        num_classes:
+            Number of classes to use for evaluation.
+        iou_thresholds:
+            IoU thresholds for evaluation.
+        max_detection_thresholds:
+            Thresholds on max detections per image.
+        class_metrics:
+            Option to enable per-class metrics for mAP and mAR_100. Has a performance impact. default: False
+        compute_on_step:
+            Forward only calls ``update()`` and return ``None`` if this is set to ``False``.
+        dist_sync_on_step:
+            Synchronize metric state across processes at each ``forward()``
+            before returning the value at the step
+        process_group:
+            Specify the process group on which synchronization is called.
+            default: ``None`` (which selects the entire world)
+        dist_sync_fn:
+            Callback that performs the allgather operation on the metric state. When ``None``, DDP
+            will be used to perform the allgather
+
+    Raises:
+        ImportError:
+            If ``torchvision`` is not installed or version installed is lower than 0.8.0
+        ValueError:
+            If ``class_metrics`` is not a boolean
+    """
+
+    def __init__(
+        self,
+        in_fmt: str = "xyxy",
+        num_classes: Optional[int] = None,
```

Comment:
In the current implementation we could avoid the num_classes and determine those with the collected `preds` and `targets`. Have a look at the `get_classes` method of my latest PR.